src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  }
+  return nullptr;
+}
+
+Node* RShiftNode::IdentityIL(PhaseGVN* phase, BasicType bt) {
   int count = 0;
   if (const_shift_count(phase, this, &count)) {
-    if ((count & (BitsPerJavaInteger - 1)) == 0) {
-      // Shift by a multiple of 32 does nothing
+    if ((count & (bits_per_java_integer(bt) - 1)) == 0) {
+      // Shift by a multiple of 32/64 does nothing
       return in(1);
     }
     // Check for useless sign-masking
-    if (in(1)->Opcode() == Op_LShiftI &&
+    if (in(1)->Opcode() == Op_LShift(bt) &&
         in(1)->req() == 3 &&
         in(1)->in(2) == in(2)) {
-      count &= BitsPerJavaInteger-1; // semantics of Java shifts
+      count &= bits_per_java_integer(bt)-1; // semantics of Java shifts
       // Compute masks for which this shifting doesn't change
-      int lo = (-1 << (BitsPerJavaInteger - ((uint)count)-1)); // FFFF8000
-      int hi = ~lo;               // 00007FFF
-      const TypeInt* t11 = phase->type(in(1)->in(1))->isa_int();
+      jlong lo = (-1 << (bits_per_java_integer(bt) - ((uint)count)-1)); // FFFF8000
+      jlong hi = ~lo;               // 00007FFF
+      const TypeInteger* t11 = phase->type(in(1)->in(1))->isa_integer(bt);
       if (t11 == nullptr) {
         return this;
       }
       // Does actual value fit inside of mask?
-      if (lo <= t11->_lo && t11->_hi <= hi) {
+      if (lo <= t11->lo_as_long() && t11->hi_as_long() <= hi) {
         return in(1)->in(1);      // Then shifting is a nop
       }
     }
   }
   return this;
 }
 
+const Type* RShiftNode::ValueIL(PhaseGVN* phase, BasicType bt) const {
+  const Type *t1 = phase->type(in(1));
+  const Type *t2 = phase->type(in(2));
+  // Either input is TOP ==> the result is TOP
+  if (t1 == Type::TOP) return Type::TOP;
+  if (t2 == Type::TOP) return Type::TOP;
+
+  // Left input is ZERO ==> the result is ZERO.
+  if (t1 == TypeInteger::zero(bt)) return TypeInteger::zero(bt);
+  // Shift by zero does nothing
+  if (t2 == TypeInt::ZERO) return t1;
+
+  // Either input is BOTTOM ==> the result is BOTTOM
+  if (t1 == Type::BOTTOM || t2 == Type::BOTTOM) {
+    return TypeInteger::bottom(bt);
+  }
+
+  const TypeInteger* r1 = t1->isa_integer(bt); // Handy access
+  const TypeInt* r2 = t2->isa_int(); // Handy access
+
+  // If the shift is a constant, just shift the bounds of the type.
+  // For example, if the shift is 31/63, we just propagate sign bits.
+  if (!r1->is_con() && r2->is_con()) {
+    uint shift = r2->get_con();
+    shift &= bits_per_java_integer(bt)-1;  // semantics of Java shifts
+    // Shift by a multiple of 32/64 does nothing:
+    if (shift == 0)  return t1;
+    // Calculate reasonably aggressive bounds for the result.
+    // This is necessary if we are to correctly type things
+    // like (x<<24>>24) == ((byte)x).
+    jlong lo = r1->lo_as_long() >> (jint)shift;
+    jlong hi = r1->hi_as_long() >> (jint)shift;
+    assert(lo <= hi, "must have valid bounds");
+#ifdef ASSERT
+   if (bt ==T_INT) {
```

Comment:
```suggestion
   if (bt == T_INT) {
```
Could this assert be generic to also handle T_LONG too?

Reply:
The assert checks that, for the int case:

```
long lo;
assert((int)(lo >> shift) == (((int)lo) >> shift, "");
```
For long, it would be:
```
long lo;
assert((long)(lo >> shift) == (((long)lo) >> shift, "");
```
Given everything is already a long, that's:
```
long lo;
assert(lo >> shift == lo >> shift, "");
```

Reply:
Ah I see, thank you for the explanation!